### ISSUE TITLE
Bug/Feature: constant_folding FP16 

### ISSUE THREAD

**Issue by KillPinguin** [tensorflow/tensorflow#15521]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
RHEL 7
- **TensorFlow installed from (source or binary)**:
Source
- **TensorFlow version (use command below)**:
('v1.3.0-rc1-6207-ge210cb1', '1.4.0')
- **Python version**: 
python2.7
- **Bazel version (if compiling from source)**:
bazel 0.9.0
- **GCC/Compiler version (if compiling from source)**:
gcc 4.8
- **Exact command to reproduce**:

python tensorflow/models/tutorial/images/convolutional.py --use_fp16

### Describe the problem
When using FP16 in e.g. the case listed above, the following error occurs:
E tensorflow/core/grappler/optimizers/constant_folding.cc:1272] Unexpected type half
E tensorflow/core/grappler/optimizers/constant_folding.cc:1242] Unexpected type half

When looking into constant_folding.cc, i found out FP16 support exists at given lines, but is commented out.
Why is this not yet included in Tensorflow?

As far as I can see, this is more of a feature request than a bug report, since these code lines simply check if computational effort can be reduced (if the matrix is zero or one)

Also when using above command, the model trains until step 1100, then the learning rate drops to 0 (a known FP16 problem). Still it's annoying to encounter this in an official tutorial file.

**Comment by jart** [comment]
@rmlarsen Any particular reason DT_HALF was left commented-out in constant_folding.cc? cl/177491247

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by rmlarsen** [comment]
@KillPinguin thank for bringing this to our attention.
@jart The code doesn't compile with fp16. We would welcome a PR adding support. In the meantime, I'll change the code just skip the optimization without spamming the error log. It is at worst a missed optimization, but certainly not an error.

**Comment by rmlarsen** [comment]
The change was submitted internally and should be visible on github within a day or so.
